gulpfile.js

```diff
@@ -5,21 +5,142 @@
+    return new debugInstall.DebugInstaller(new debugUtil.CoreClrDebugUtil(path.resolve('.')));
+}
+
+function installDebugger(runtimeId) {
+    return getDebugInstaller().install(runtimeId);
+}
+
+function cleanDebugger() {
+    try {
+        getDebugInstaller().clean();
+        console.log('Cleaned Succesfully');
+    } catch (error) {
+        console.error(error);
+    }
+}
+
+gulp.task('debugger:install', ['debugger:clean'], () => {
+    installDebugger(gulp.env.runtimeId).then(() => {
+        console.log('Installed Succesfully');
+    }).catch((error) => {
+        console.error(error);
+    });
+});
+
+gulp.task('debugger:clean', () => {
+    cleanDebugger();
+});
+
+/// Packaging Tasks
+function doPackage(packageName) {
+
+    var vsceArgs = [];
+    vsceArgs.push(path.join(__dirname, 'node_modules', 'vsce', 'out', 'vsce'))
+    vsceArgs.push('package'); // package command
+
+    if (packageName !== undefined) {
+        vsceArgs.push('-o');
+        vsceArgs.push(packageName);
+    }
+
+    var proc = child_process.spawnSync('node', vsceArgs);
+    if (proc.error) {
+        console.error(proc.error.toString());
+    }
+}
+
+function doOfflinePackage(runtimeId, omnisharpAsset, packageName) {
+    return clean().then(function() {
+        return installDebugger(runtimeId);
+    }).then(function() {
+        return installOmnisharp(omnisharpAsset);
+    }).then(function () {
+        doPackage(packageName + '-' + runtimeId + '.vsix');
+    });
+}
+
+gulp.task('package:clean', () => {
+    return del('*.vsix');
+});
+
+gulp.task('package:online', ['clean'], () => {
+    doPackage();
 });
 
-//TODO: decouple omnisharpDownload (specifically proxy.ts) from vscode 
-// gulp.task('omnisharp:fetch', ['omnisharp:clean'], () => {
-// 	return omnisharpDownload.downloadOmnisharp();
-// });
+gulp.task('package:offline', ['clean'], () => {
+    var json = JSON.parse(fs.readFileSync('package.json'));
+    var name = json.name;
+    var version = json.version;
+    var packageName = name + '.' + version;
 
+    var packages = [];
+    packages.push({rid: 'win7-x64', omni: `omnisharp-${OmniSharpVersion}-win-x64-net451.zip`});
+    packages.push({rid: 'osx.10.11-x64', omni: `omnisharp-${OmniSharpVersion}-osx-x64-netcoreapp1.0.tar.gz`});
+    packages.push({rid: 'centos.7-x64', omni: `omnisharp-${OmniSharpVersion}-centos-x64-netcoreapp1.0.tar.gz`});
+    packages.push({rid: 'debian.8-x64', omni: `omnisharp-${OmniSharpVersion}-debian-x64-netcoreapp1.0.tar.gz`});
+    packages.push({rid: 'rhel.7.2-x64', omni: `omnisharp-${OmniSharpVersion}-rhel-x64-netcoreapp1.0.tar.gz`});
+    packages.push({rid: 'ubuntu.14.04-x64', omni: `omnisharp-${OmniSharpVersion}-ubuntu-x64-netcoreapp1.0.tar.gz`});
+
+    var promise = Promise.resolve();
+
+    packages.forEach(function(pair) {
+        promise = promise.then(function() {
+            return doOfflinePackage(pair.rid, pair.omni, packageName);
+        })
+    });
+
+    return promise;
```

Comment:
I think I had tried that and it caused issue because all those promises will execute concurrently, while we __need__ them to execute synchronously.